Request for comments: deciding upon themes in ECL v3 (#1703)
ec-europa/europa-component-library: docs/decisions/005-themes.md

## Scopes

The topics below are categorized with focus on separation of subjects rather than specific type of code.

### Package organization

Although the concept of "presets" was put under question during the analysis of v3, they are still used throughout this document in terms of code organization, not ECL release packages. This means that regardless of the usage of presets below, they may not be part of the resulting release of ECL v3.

#### Option 1: ECL v2 is preserved

This option has been discarded from the very beginning because it cannot achieve the goal without increased complexity and code duplication.

#### Option 2: "themes" is single package, single preset generates multiple themes output

This option has been discarded as too complex for the worth of the result. The complexity comes from SASS language limitations as dynamic imports are not supported. Every dynamic theme generation approach with SASS involves workarounds which were not acceptable for the team.

#### Option 3: "themes" is single package, multiple presets generate multiple themes output

> - `@ecl/ecl-base` - common resources

Reviewer: ```suggestion
- `@ecl/ecl-base` - common variables
```

> - `@ecl/ecl-theme` - set of variables, one per theme

Reviewer: ```suggestion
- `@ecl/ecl-theme` - set of variables related to the themes (global parameters), one set per theme
```

- `@ecl/ecl-preset-{ec-core}` - base + theme ec core + components
- `@ecl/ecl-preset-{eu-core}` - base + theme eu core + components
- `@ecl/ecl-preset-{ec-standardised}` - base + theme ec standardised + components
- `@ecl/ecl-preset-{eu-standardised}` - base + theme eu standardised + components
- `@ecl/ecl-preset-{custom}` - template for theme generation

#### Option 4: "themes" are multiple packages, multiple presets generate multiple themes output

Same as option 3 with the following difference in theme-related packages which are multiple:

- `@ecl/ecl-theme-{ec-core}` - set of variables
- `@ecl/ecl-theme-{eu-core}` - set of variables
- `@ecl/ecl-theme-{ec-standardised}` - set of variables
- `@ecl/ecl-theme-{eu-standardised}` - set of variables

In this version, each target bundle is composed of a pair of a theme + preset.

### Resource organization

Resources are: logo, favicons, icons and other similar types of assets which are not SCSS or JavaScript and are used by consumers of ECL.

#### Option 1: separate package per system or per theme

Current organization of v2.

- `@ecl/ec-resources-logo`
- `@ecl/eu-resources-logo`

> #### Option 2: single package

Reviewer: That's what we had in the begining, but we split it in several packages to allow users to grab just what they need

Author: Didn't know. Please do add this info similarly to https://github.com/ec-europa/europa-component-library/pull/1703/files#diff-25588a9df5ef0f08c630d3d1e730dffdR42

> - `@ecl/resources`

Reviewer: ```suggestion
- `@ecl/resources`

#### Option 3: include resources in themes

This option does imply that each theme comes in a dedicates package (see ahead).
Resources useful to display a theme are stored directly in it.
To limit duplication, some global resources may still be put in dedicated packages.
```

# Handle `themes` in ECL v3

| Status        | work in progress                                      |
| ------------- | ----------------------------------------------------- |
| **Proposed**  | 07/10/2020                                            |
| **Accepted**  | --/--/---                                             |
| **Driver**    | @kalinchernev                                         |
| **Approver**  |                                                       |
| **Consulted** | @emeryro                                              |
| **Informed**  | [@team](https://github.com/orgs/ec-europa/teams/inno) |

## Decisions

## Context

> - ECL v2 is based on the concept of `systems` such as EC and EU

Reviewer 2: ```suggestion
- ECL v2 was made in order to generate two independent design systems, EC and EU.
```

> - ECL v3 needs to accommodate the concept of `themes`.

Reviewer 2: ```suggestion
- ECL V3 starts from the assumption that there is only one source template and one design system with multiple `themes`.
```
